sw/device/lib/testing/mock_mmio.h

```diff
@@ -0,0 +1,215 @@
+ *
+ * This class should be derived by a test fixture (along with |testing::Test|)
+ * and used in a |TEST_F| block. Doing so will make the |EXPECT_READN| and
+ * |EXPECT_WRITEN| conveinence macros useable.
+ * 
+ * The device being mocked can be accessed in the test body with |this->dev()|.
+ * |this->| is required in this case, since the name |dev| is not immediately
+ * visible.
+ */
+class MmioTest {
+ protected:
+  MockDevice &dev() { return *dev_; }
+ private:
+  std::unique_ptr<MockDevice> dev_ = std::make_unique<MockDevice>();
+};
+
+}  // namespace mock_mmio
+
+/**
+ * Expect a read to the device |dev| at the given offset, returning the given
+ * 8-bit value.
+ *
+ * This expectation is sequenced with all other |EXPECT_READ| and |EXPECT_WRITE|
+ * calls.
+ */
+#define EXPECT_READ8_AT(dev, offset, value)  \
+  EXPECT_CALL(dev, Read8(offset)) \
+      .InSequence(dev.seq_)              \
+      .WillOnce(testing::Return(value))
+
+/**
+ * Expect a read to the device |dev| at the given offset, returning the given
+ * 16-bit value.
+ *
+ * This expectation is sequenced with all other |EXPECT_READ| and |EXPECT_WRITE|
+ * calls.
+ */
+#define EXPECT_READ16_AT(dev, offset, value)  \
+  EXPECT_CALL(dev, Read16(offset)) \
+      .InSequence(dev.seq_)              \
+      .WillOnce(testing::Return(value))
+
+/**
+ * Expect a read to the device |dev| at the given offset, returning the given
+ * 32-bit value.
+ *
+ * This expectation is sequenced with all other |EXPECT_READ| and |EXPECT_WRITE|
+ * calls.
+ */
+#define EXPECT_READ32_AT(dev, offset, value)  \
+  EXPECT_CALL(dev, Read32(offset)) \
+      .InSequence(dev.seq_)              \
+      .WillOnce(testing::Return(value))
+
+/**
+ * Expect a write to the device |dev| at the given offset with the given 8-bit
+ * value.
+ *
+ * This expectation is sequenced with all other |EXPECT_READ| and |EXPECT_WRITE|
+ * calls.
+ */
+#define EXPECT_WRITE8_AT(dev, offset, value) \
+  EXPECT_CALL(dev, Write8(offset, value)).InSequence(dev.seq_)
+
+/**
+ * Expect a write to the device |dev| at the given offset with the given 16-bit
+ * value.
+ *
+ * This expectation is sequenced with all other |EXPECT_READ| and |EXPECT_WRITE|
+ * calls.
+ */
+#define EXPECT_WRITE16_AT(dev, offset, value) \
+  EXPECT_CALL(dev, Write16(offset, value)).InSequence(dev.seq_)
+
+/**
+ * Expect a write to the device |dev| at the given offset with the given 32-bit
+ * value.
+ *
+ * This expectation is sequenced with all other |EXPECT_READ| and |EXPECT_WRITE|
+ * calls.
+ */
+#define EXPECT_WRITE32_AT(dev, offset, value) \
+  EXPECT_CALL(dev, Write32(offset, value)).InSequence(dev.seq_)
+
+/**
+ * Expect a read to the at the given offset, returning the given 8-bit value.
+ * 
+ * This function is only available in tests using a fixture that derives |DeviceTest|.
+ *
+ * This expectation is sequenced with all other |EXPECT_READ| and |EXPECT_WRITE|
+ * calls.
+ */
+#define EXPECT_READ8(offset, value) \
```

Comment:
I don't quite understand why we need two sets of these macros. In this change there is only example of usage for EXPECT_<READn, WRITEn> macros. Could you please explicitly document how these are different from ...AT macros? Is there need for two set of macros?

Reply:
The `_AT` versions expect you to provide the mmio_region value, while these versions use the implicit one implied by the test fixture. The former exist to be used in helper functions that need to be outside of the test fixture's scope.

Reply:
It's not immediately obvious from this change how the _AT macros are supposed to be used, maybe it is worth expanding the description?